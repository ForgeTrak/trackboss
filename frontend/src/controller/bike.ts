import { generateHeaders } from './utils';
import {
    DeleteBikeResponse,
    GetBikeListResponse,
    GetBikeResponse,
    PatchBikeRequest,
    PatchBikeResponse,
    PostNewBikeRequest,
    PostNewBikeResponse,
} from '../../../src/typedefs/bike';

export async function createBike(token: string, bikeData: PostNewBikeRequest): Promise<PostNewBikeResponse> {
    const response = await fetch(`${process.env.REACT_APP_API_URL}/api/bike/new`, {
        method: 'POST',
        mode: 'no-cors',
        headers: generateHeaders(token),
        body: JSON.stringify(bikeData),
    });
    return response.json();
}

export async function getBikeList(token: string, membershipID?: number): Promise<GetBikeListResponse> {
    if (membershipID) {
        const idString = membershipID.toString();
        const response = await fetch(`${process.env.REACT_APP_API_URL}/api/bike/list?membershipID=${idString}`, {
            method: 'GET',
            mode: 'cors',
            headers: generateHeaders(token),
        });
        return response.json();
    }
    // else
    const response = await fetch(`${process.env.REACT_APP_API_URL}/api/bike/list`, {
        method: 'GET',
        mode: 'cors',
        headers: generateHeaders(token),
    });
    return response.json();
}

export async function getBike(token: string, bikeID: number): Promise<GetBikeResponse> {
    const response = await fetch(`${process.env.REACT_APP_API_URL}/api/bike/${bikeID}`, {
        method: 'GET',
        mode: 'no-cors',
        headers: generateHeaders(token),
    });
    return response.json();
}

export async function updateBike(
    token: string,
    bikeID: number,
    bikeData: PatchBikeRequest,
): Promise<PatchBikeResponse> {
    const response = await fetch(`${process.env.REACT_APP_API_URL}/api/bike/${bikeID}`, {
        method: 'PATCH',
        mode: 'no-cors',
        headers: generateHeaders(token),
        body: JSON.stringify(bikeData),
    });
    return response.json();
}

export async function deleteBike(token: string, bikeID: number): Promise<DeleteBikeResponse> {
    const response = await fetch(`${process.env.REACT_APP_API_URL}/api/bike/${bikeID}`, {
        method: 'DELETE',
        mode: 'no-cors',
        headers: generateHeaders(token),
    });
    return response.json();
}
