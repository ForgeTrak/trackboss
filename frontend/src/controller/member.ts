import generateHeaders from "./utils";


export async function createMember(token: string, memberData: object) {
    const response = await fetch(`${process.env.REACT_APP_API_URL}/api/member/new`, {
        method: 'POST',
        mode: 'no-cors',
        headers: generateHeaders(token),
        body: JSON.stringify(memberData),
    });
    return response.json();
}

export async function getMember(token: string, memberId: number) {
    const response = await fetch(`${process.env.REACT_APP_API_URL}/api/member/${memberId}`, {
        method: 'GET',
        mode: 'no-cors',
        headers: generateHeaders(token),
    });
    return response.json();
}

export async function getMemberList(token: string, listType?: string) {
    if (listType) {
        const response = await fetch(`${process.env.REACT_APP_API_URL}/api/member/list?status=${listType}`, {
            method: 'GET',
            mode: 'no-cors',
            headers: generateHeaders(token),
        });
        return response.json();
    }
    // else
    const response = await fetch(`${process.env.REACT_APP_API_URL}/api/member/list`, {
        method: 'GET',
        mode: 'no-cors',
        headers: generateHeaders(token),
    });
    return response.json();
}

export async function updateMember(token: string, memberID: number, memberData: object) {
    const response = await fetch(`${process.env.REACT_APP_API_URL}/api/member/${memberID}`, {
        method: 'PATCH',
        mode: 'no-cors',
        headers: generateHeaders(token),
        body: JSON.stringify(memberData),
    });
    return response.json();
}