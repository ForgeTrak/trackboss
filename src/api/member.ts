import { Request, Response, Router } from 'express';

const member = Router();

member.post('/new', (req: Request, res: Response) => {
    res.status(501).send();
});

member.get('/list', (req: Request, res: Response) => {
    res.status(501).send();
});

member.get('/:memberID', (req: Request, res: Response) => {
    res.status(501).send();
});

member.patch('/:memberID', (req: Request, res: Response) => {
    res.status(501).send();
});

member.delete('/:memberID', (req: Request, res: Response) => {
    res.status(501).send();
});

export default member;
