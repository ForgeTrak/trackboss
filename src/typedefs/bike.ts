import { ErrorResponse } from './errorResponse';

export type Bike = {
    bikeId: number,
    year: string,
    make: string,
    model: string,
    membershipAdmin: string
}

export type DeletedBike = {
    bikeId: number
}

export type PostNewBikeRequest = {
    year?: string,
    make?: string,
    model?: string,
    membershipId: number
}

export type PostNewBikeResponse = Bike | ErrorResponse

export type GetBikeRequest = Record<string, never>

export type GetBikeResponse = Bike | ErrorResponse

export type PatchBikeRequest = {
    year?: string,
    make?: string,
    model?: string,
    membershipId?: number
}

export type PatchBikeResponse = Bike | ErrorResponse

export type DeleteBikeRequest = Record<string, never>

export type DeleteBikeResponse = DeletedBike | ErrorResponse

export type GetBikeListRequest = Record<string, never>

export type GetBikeListResponse = Bike[] | ErrorResponse
