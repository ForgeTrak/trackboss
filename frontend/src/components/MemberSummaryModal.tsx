/* eslint-disable no-unused-vars */
/* eslint-disable max-len */
import React, { useEffect, useState } from 'react';
import {
    Modal,
    ModalOverlay,
    ModalContent,
    ModalHeader,
    ModalBody,
    ModalCloseButton,
    Button,
    useDisclosure,
    Text,
    SimpleGrid,
    Heading,
    VStack,
    Divider,
    Container,
    HStack,
    UnorderedList,
    ListItem,
    ButtonGroup,
    Input,
    Select,
} from '@chakra-ui/react';
import { Member } from '../../../src/typedefs/member';
import { Bike } from '../../../src/typedefs/bike';

interface modalProps {
    memberInfo: Member,
    memberFamily: Member[],
    memberBikes: Bike[],
    // admin: boolean, // TODO: this will come from state
}

async function handleNewBoardMember(memberInfo: Member, editedBoardMember: string) {
    // const currentYear = new Date().getFullYear();
    // const boardRoleId = getBoardRoleId(editedBoardMember); // Util method?
    // const response = await newBoardMember(currentYear,boardRoleId, memberInfo.memberId);
    // return response;
}

async function handlePatchMemberType(memberInfo: Member, editedMemberType: string) {
    // const updatedMember = await patchMember(editedMemberType, memberInfo.memberId)
    // if (updatedMember.reason) {
    //   there was an error, show error message
    // }
}

async function handlePatchMemberContactInfo(
    memberInfo: Member,
    name: string | undefined,
    email: string | undefined,
    phone: string | undefined,
) {
    // const updatedMember = await patchMember(name, email, phone, memberInfo.memberId)
    // if (updatedMember.reason) {
    //      there was an error, show error message
    // }

    const updatedMember: Member = {
        memberId: 1,
        membershipAdmin: 'true',
        active: true,
        memberType: 'member',
        firstName: '',
        lastName: 'gottem',
        phoneNumber: '0987',
        email: 'gottem@example.com',
        uuid: '',
        occupation: '',
        birthdate: '',
        dateJoined: '',
        address: '',
        city: '',
        state: '',
        zip: '',
        lastModifiedDate: '',
        lastModifiedBy: '',
    };

    return updatedMember;
}

export default function MemberSummaryModal(props: modalProps) {
    const { isOpen, onOpen, onClose } = useDisclosure();
    const [memberInfo, setMemberInfo] = useState<Member>();
    // const [family, setFamily] = useState<Member[]>();
    const [bikes, setBikes] = useState<Bike[]>();

    const [editingMemberInfo, setEditingMemberInfo] = useState<boolean>();
    const [editingMemberRole, setEditingMemberRole] = useState<boolean>();

    const [editedName, setEditedName] = useState<string>('');
    const [editedEmail, setEditedEmail] = useState<string>('');
    const [editedPhone, setEditedPhone] = useState<string>('');

    const [editedMemberType, setEditedMemberType] = useState<string>('');
    const [editedBoardMember, setEditedBoardMember] = useState<string>('');

    const handleEditedNameChange = (event: { target: { value: any; }; }) => setEditedName(event.target.value);
    const handleEditedEmailChange = (event: { target: { value: any; }; }) => setEditedEmail(event.target.value);
    const handleEditedPhoneChange = (event: { target: { value: any; }; }) => setEditedPhone(event.target.value);
    const handleEditedBoardMember = (event: { target: { value: any; }; }) => setEditedBoardMember(event.target.value);

    useEffect(() => {
        async function setModalData() {
            setMemberInfo(props.memberInfo);
            setEditedMemberType(props.memberInfo.memberType);
            // setFamily(props.memberFamily);
            setBikes(props.memberBikes);
            setEditingMemberInfo(false);
            setEditingMemberRole(false);
        }
        setModalData();
    }, []);

    return (
        <div>
            <Button background="orange.300" color="white" onClick={onOpen}>Member Summary Modal</Button>
            <Modal size="xl" isOpen={isOpen} onClose={onClose}>
                <ModalOverlay />
                <ModalContent pb={5}>
                    <ModalHeader><Heading textAlign="center">Member Summary</Heading></ModalHeader>
                    <Divider />
                    <ModalCloseButton />
                    <ModalBody>
                        {
                            memberInfo && bikes && (
                                <SimpleGrid columns={2} spacing={4}>
                                    <VStack borderRightWidth={1} borderRightColor="light-grey" align="left">
                                        <HStack>
                                            <Text textAlign="left" fontSize="3xl" fontWeight="bold">Contact Info</Text>
                                            {
                                                memberInfo.membershipAdmin === 'true' && (
                                                    <Button
                                                        textDecoration="underline"
                                                        color="orange"
                                                        variant="ghost"
                                                        onClick={
                                                            () => {
                                                                if (editingMemberInfo) {
                                                                    setEditingMemberInfo(false);
                                                                } else {
                                                                    setEditingMemberInfo(true);
                                                                }
                                                            }
                                                        }
                                                    >
                                                        Edit
                                                    </Button>
                                                )
                                            }
                                        </HStack>
                                        <SimpleGrid pb={4} columns={2}>
                                            <VStack maxWidth={40} spacing={2} align="left">
                                                <Text fontWeight="bold">Name:</Text>
                                                <Text fontWeight="bold">Email:</Text>
                                                <Text fontWeight="bold">Phone:</Text>
                                            </VStack>
                                            {
                                                editingMemberInfo ? (
                                                    <VStack ml="-65px" align="left">
                                                        <Input
                                                            placeholder={`${memberInfo.firstName} ${props.memberInfo.lastName}`}
                                                            value={editedName}
                                                            onChange={handleEditedNameChange}
                                                            size="xs"
                                                        />
                                                        <Input
                                                            placeholder={memberInfo.email}
                                                            value={editedEmail}
                                                            onChange={handleEditedEmailChange}
                                                            size="xs"
                                                        />
                                                        <Input
                                                            placeholder={memberInfo.phoneNumber}
                                                            value={editedPhone}
                                                            onChange={handleEditedPhoneChange}
                                                            size="xs"
                                                        />
                                                        <Button
                                                            ml={10}
                                                            variant="outline"
                                                            size="xs"
                                                            color="green"
                                                            onClick={
                                                                async () => {
                                                                    setMemberInfo(await handlePatchMemberContactInfo(memberInfo, editedName, editedEmail, editedPhone));
                                                                    setEditingMemberInfo(false);
                                                                    setEditingMemberRole(false);
                                                                }
                                                            }
                                                        >
                                                            Save
                                                        </Button>
                                                    </VStack>
                                                ) : (
                                                    <VStack ml="-65px" align="left">
                                                        <Text>
                                                            {`${memberInfo.firstName} ${props.memberInfo.lastName}`}
                                                        </Text>
                                                        <Text>{memberInfo.email}</Text>
                                                        <Text>{memberInfo.phoneNumber}</Text>
                                                    </VStack>
                                                )
                                            }
                                        </SimpleGrid>
                                        <HStack>
                                            <Text textAlign="left" fontSize="3xl" fontWeight="bold">Family</Text>
                                            {/* {
                                                memberInfo.membershipAdmin === 'true' && (
                                                    <Button
                                                        textDecoration="underline"
                                                        color="orange"
                                                        variant="ghost"
                                                    >
                                                        Edit
                                                    </Button>
                                                )
                                            } */}
                                        </HStack>
                                        <Text textAlign="left" fontSize="1xl" fontWeight="bold">Members</Text>
                                        <UnorderedList pl={10}>
                                            {
                                                props.memberFamily.map((member) => (
                                                    member.memberId === memberInfo.memberId ? <ListItem>{`${member.firstName} ${member.lastName} (you)`}</ListItem>
                                                        : <ListItem>{`${member.firstName} ${member.lastName}`}</ListItem>
                                                ))
                                            }
                                        </UnorderedList>
                                        <Text textAlign="left" fontSize="1xl" fontWeight="bold">Bikes</Text>
                                        <UnorderedList pl={10}>
                                            {
                                                props.memberBikes.map((bike) => (
                                                    <ListItem>{`${bike.year}, ${bike.make} ${bike.model}`}</ListItem>
                                                ))
                                            }
                                        </UnorderedList>
                                    </VStack>
                                    <VStack align="left">
                                        <HStack>
                                            <Text
                                                textAlign="left"
                                                fontSize="3xl"
                                                fontWeight="bold"
                                            >
                                                Roles
                                            </Text>
                                            {
                                                memberInfo.membershipAdmin === 'true' && (
                                                    <Button
                                                        textDecoration="underline"
                                                        color="orange"
                                                        variant="ghost"
                                                        onClick={
                                                            () => {
                                                                if (editingMemberRole) {
                                                                    setEditingMemberRole(false);
                                                                } else {
                                                                    setEditingMemberRole(true);
                                                                }
                                                            }
                                                        }
                                                    >
                                                        Edit
                                                    </Button>
                                                )
                                            }
                                        </HStack>
                                        {
                                            editingMemberRole ? (
                                                <VStack align="left">
                                                    <ButtonGroup size="sm" isAttached variant="outline">
                                                        <Button
                                                            onClick={
                                                                async () => {
                                                                    if (editedMemberType !== 'member') {
                                                                        await setEditedMemberType('member');
                                                                    }
                                                                }
                                                            }
                                                            mr="-px"
                                                            backgroundColor={editedMemberType === 'member' ? 'blue' : ''}
                                                        >
                                                            Member
                                                        </Button>
                                                        <Button
                                                            onClick={
                                                                async () => {
                                                                    if (editedMemberType !== 'admin') {
                                                                        await setEditedMemberType('admin');
                                                                    }
                                                                }
                                                            }
                                                            mr="-px"
                                                            backgroundColor={editedMemberType === 'admin' ? 'blue' : ''}
                                                        >
                                                            Admin
                                                        </Button>
                                                        <Button
                                                            onClick={
                                                                async () => {
                                                                    if (editedMemberType !== 'board') {
                                                                        await setEditedMemberType('board');
                                                                    }
                                                                }
                                                            }
                                                            mr="-px"
                                                            backgroundColor={editedMemberType === 'board' ? 'blue' : ''}
                                                        >
                                                            Board
                                                        </Button>
                                                    </ButtonGroup>
                                                    <Select
                                                        disabled={editedMemberType !== 'board'}
                                                        variant="outline"
                                                        size="xs"
                                                        placeholder={memberInfo.occupation}
                                                        value={editedBoardMember}
                                                        onChange={handleEditedBoardMember}
                                                    >
                                                        {/* TODO: need the actual roles here */}
                                                        <option value="president">president</option>
                                                        <option value="vice president">vice president</option>
                                                        <option value="secretary">secretary</option>
                                                    </Select>
                                                    <Button
                                                        ml={10}
                                                        variant="outline"
                                                        size="xs"
                                                        color="green"
                                                        onClick={
                                                            async () => {
                                                                if (editedBoardMember !== '') {
                                                                    await handleNewBoardMember(memberInfo, editedBoardMember);
                                                                }
                                                                if (memberInfo.memberType !== editedMemberType) {
                                                                    await handlePatchMemberType(memberInfo, editedMemberType);
                                                                }
                                                                setEditingMemberRole(false);
                                                                setEditingMemberInfo(false);
                                                            }
                                                        }
                                                    >
                                                        Save
                                                    </Button>
                                                </VStack>
                                            ) : (
                                                <ButtonGroup size="sm" isAttached variant="outline">
                                                    <Button
                                                        mr="-px"
                                                        backgroundColor={memberInfo.memberType === 'member' ? 'blue' : ''}
                                                    >
                                                        Member
                                                    </Button>
                                                    <Button
                                                        userSelect="none"
                                                        mr="-px"
                                                        backgroundColor={memberInfo.memberType === 'admin' ? 'blue' : ''}
                                                    >
                                                        Admin
                                                    </Button>
                                                    <Button
                                                        mr="-px"
                                                        backgroundColor={memberInfo.memberType === 'board' ? 'blue' : ''}
                                                    >
                                                        Board
                                                    </Button>
                                                </ButtonGroup>
                                            )
                                        }
                                        <Container pt={20} mb={20} />
                                        <Text textAlign="left" fontSize="3xl" fontWeight="bold">Actions</Text>
                                        <HStack size="lg">
                                            <Button
                                                variant="outline"
                                                style={
                                                    {
                                                        whiteSpace: 'normal',
                                                        wordWrap: 'break-word',
                                                    }
                                                }
                                            >
                                                Assign To Event
                                            </Button>
                                            <Button
                                                backgroundColor="red"
                                                variant="outline"
                                                style={
                                                    {
                                                        whiteSpace: 'normal',
                                                        wordWrap: 'break-word',
                                                    }
                                                }
                                            >
                                                De-Activate Member
                                            </Button>
                                        </HStack>
                                        <Button
                                            textDecoration="underline"
                                            color="orange"
                                            variant="ghost"
                                        >
                                            Reset Password
                                        </Button>
                                    </VStack>
                                </SimpleGrid>
                            )
                        }
                    </ModalBody>
                </ModalContent>
            </Modal>
        </div>
    );
}
