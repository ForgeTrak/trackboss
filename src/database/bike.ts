import _ from 'lodash';
import { OkPacket, RowDataPacket } from 'mysql2';

import logger from '../logger';
import pool from './pool';
import { Bike, PostNewBikeRequest } from '../typedefs/bike';

export const INSERT_BIKE_SQL = 'INSERT INTO member_bikes (year, make, model, membership_id) VALUES (?, ?, ?, ?)';
export const GET_BIKE_SQL = 'SELECT bike_id, year, make, model, membership_admin FROM v_bike WHERE bike_id = ?';

export async function insertBike(req: PostNewBikeRequest): Promise<number> {
    //const sql = 'INSERT INTO member_bikes (year, make, model, membership_id) VALUES (?, ?, ?, ?)';
    const values = [req.year, req.make, req.model, req.membershipId];

    let result;
    try {
        [result] = await pool.query<OkPacket>(INSERT_BIKE_SQL, values);
    } catch (e) {
        logger.error(`DB error inserting bike: ${e}`);
        throw new Error('internal server error');
    }

    return result.insertId;
}

export async function getBike(id: number): Promise<Bike> {
    //const sql = 'SELECT bike_id, year, make, model, membership_admin FROM v_bike WHERE bike_id = ?';
    const values = [id];

    let results;
    try {
        [results] = await pool.query<RowDataPacket[]>(GET_BIKE_SQL, values);
    } catch (e) {
        logger.error(`DB error getting bike: ${e}`);
        throw new Error('internal server error');
    }

    if (_.isEmpty(results)) {
        throw new Error('not found');
    }

    return {
        bikeId: results[0].bike_id,
        year: results[0].year,
        make: results[0].make,
        model: results[0].model,
        membershipAdmin: results[0].membership_admin,
    };
}
