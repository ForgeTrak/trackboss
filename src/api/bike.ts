import { Request, Response, Router } from 'express';
import { deleteBike, getBike, getBikeList, insertBike, patchBike } from '../database/bike';
import {
    Bike,
    DeleteBikeResponse,
    GetBikeListResponse,
    GetBikeResponse,
    PatchBikeResponse,
    PostNewBikeResponse,
} from '../typedefs/bike';
import { checkHeader, verify } from '../util/auth';

const bike = Router();

bike.post('/new', async (req: Request, res: Response) => {
    const { authorization } = req.headers;
    let response: PostNewBikeResponse;
    const headerCheck = checkHeader(authorization);
    if (!headerCheck.valid) {
        res.status(401);
        response = { reason: headerCheck.reason };
    } else {
        try {
            await verify(headerCheck.token, 'Membership Admin');
            const insertId = await insertBike(req.body);
            response = await getBike(insertId);
            res.status(201);
        } catch (e: any) {
            if (e.message === 'user input error') {
                res.status(400);
                response = { reason: 'bad request' };
            } else if (e.message === 'Authorization Failed') {
                res.status(401);
                response = { reason: 'not authorized' };
            } else if (e.message === 'Forbidden') {
                res.status(403);
                response = { reason: 'forbidden' };
            } else {
                res.status(500);
                response = { reason: 'internal server error' };
            }
        }
    }
    res.send(response);
});

bike.get('/list', async (req: Request, res: Response) => {
    const { authorization } = req.headers;
    let response: GetBikeListResponse;
    const headerCheck = checkHeader(authorization);
    if (!headerCheck.valid) {
        res.status(401);
        response = { reason: headerCheck.reason };
    } else {
        try {
            await verify(headerCheck.token);
            const id = req.query.membershipID;
            let filterMembership: number | undefined;
            if (typeof id === 'undefined') {
                filterMembership = undefined;
            } else {
                filterMembership = Number(id);
                if (Number.isNaN(filterMembership)) {
                    throw new Error('user input error');
                }
            }
            const bikeList: Bike[] = await getBikeList(filterMembership);
            res.status(200);
            response = bikeList;
        } catch (e: any) {
            if (e.message === 'user input error') {
                res.status(400);
                response = { reason: 'bad request' };
            } else if (e.message === 'Authorization Failed') {
                res.status(401);
                response = { reason: 'not authorized' };
            } else {
                res.status(500);
                response = { reason: 'internal server error' };
            }
        }
    }
    res.send(response);
});

bike.get('/:bikeID', async (req: Request, res: Response) => {
    const { authorization } = req.headers;
    let response: GetBikeResponse;
    const headerCheck = checkHeader(authorization);
    if (!headerCheck.valid) {
        res.status(401);
        response = { reason: headerCheck.reason };
    } else {
        try {
            await verify(headerCheck.token);
            const { bikeID } = req.params;
            response = await getBike(Number(bikeID));
            res.status(200);
        } catch (e: any) {
            if (e.message === 'not found') {
                res.status(404);
                response = { reason: 'not found' };
            } else if (e.message === 'Authorization Failed') {
                res.status(401);
                response = { reason: 'not authorized' };
            } else {
                res.status(500);
                response = { reason: 'internal server error' };
            }
        }
    }
    res.send(response);
});

bike.patch('/:bikeID', async (req: Request, res: Response) => {
    const { authorization } = req.headers;
    let response: PatchBikeResponse;
    const headerCheck = checkHeader(authorization);
    if (!headerCheck.valid) {
        res.status(401);
        response = { reason: headerCheck.reason };
    } else {
        try {
            const { bikeID } = req.params;
            const bikeIdNum = Number(bikeID);
            if (Number.isNaN(bikeIdNum)) {
                throw new Error('not found');
            }
            await verify(headerCheck.token, 'Membership Admin');
            await patchBike(bikeIdNum, req.body);
            response = await getBike(bikeIdNum);
            res.status(200);
        } catch (e: any) {
            if (e.message === 'user input error') {
                res.status(400);
                response = { reason: 'bad request' };
            } else if (e.message === 'not found') {
                res.status(404);
                response = { reason: 'not found' };
            } else if (e.message === 'Authorization Failed') {
                res.status(401);
                response = { reason: 'not authorized' };
            } else if (e.message === 'Forbidden') {
                res.status(403);
                response = { reason: 'forbidden' };
            } else {
                res.status(500);
                response = { reason: 'internal server error' };
            }
        }
    }
    res.send(response);
});

bike.delete('/:bikeID', async (req: Request, res: Response) => {
    const { authorization } = req.headers;
    let response: DeleteBikeResponse;
    const headerCheck = checkHeader(authorization);
    if (!headerCheck.valid) {
        res.status(401);
        response = { reason: headerCheck.reason };
    } else {
        try {
            const { bikeID } = req.params;
            const bikeIdNum = Number(bikeID);
            if (Number.isNaN(bikeIdNum)) {
                throw new Error('not found');
            }
            await verify(headerCheck.token, 'Membership Admin');
            await deleteBike(bikeIdNum);
            response = { bikeId: bikeIdNum };
            res.status(200);
        } catch (e: any) {
            if (e.message === 'not found') {
                res.status(404);
                response = { reason: 'not found' };
            } else if (e.message === 'Authorization Failed') {
                res.status(401);
                response = { reason: 'not authorized' };
            } else if (e.message === 'Forbidden') {
                res.status(403);
                response = { reason: 'forbidden' };
            } else {
                res.status(500);
                response = { reason: 'internal server error' };
            }
        }
    }
    res.send(response);
});

export default bike;
