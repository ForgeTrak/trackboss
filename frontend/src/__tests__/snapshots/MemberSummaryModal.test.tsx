import React from 'react';
import renderer from 'react-test-renderer';
import '@testing-library/jest-dom/extend-expect';
import MemberSummaryModal from '../../components/MemberSummaryModal';
import { Member } from '../../../../src/typedefs/member';
import { Bike } from '../../../../src/typedefs/bike';

const member: Member = {
    memberId: 1,
    membershipAdmin: 'true',
    active: true,
    memberType: 'admin',
    firstName: 'test',
    lastName: 'member',
    phoneNumber: '1234',
    email: 'user@example.com',
    uuid: '',
    occupation: 'secretary',
    birthdate: '',
    dateJoined: '',
    address: '',
    city: '',
    state: '',
    zip: '',
    lastModifiedDate: '',
    lastModifiedBy: '',
};

const memberFamily: Member[] = [
    {
        memberId: 1,
        membershipAdmin: 'true',
        active: true,
        memberType: 'member',
        firstName: 'test',
        lastName: 'member',
        phoneNumber: '1234',
        email: 'user@example.com',
        uuid: '',
        occupation: '',
        birthdate: '',
        dateJoined: '',
        address: '',
        city: '',
        state: '',
        zip: '',
        lastModifiedDate: '',
        lastModifiedBy: '',
    },
    {
        memberId: 2,
        membershipAdmin: 'string',
        active: true,
        memberType: 'member',
        firstName: 'Jon',
        lastName: 'Smith',
        phoneNumber: '1234',
        email: 'user@example.com',
        uuid: '',
        occupation: '',
        birthdate: '',
        dateJoined: '',
        address: '',
        city: '',
        state: '',
        zip: '',
        lastModifiedDate: '',
        lastModifiedBy: '',
    },
];

const memberBikes: Bike[] = [
    {
        bikeId: 0,
        year: '2012',
        make: 'honda',
        model: 'shitbox',
        membershipAdmin: 'string',
    },
];

describe('member summary modal', () => {
    it('renders all props correctly', () => {
        const modal = renderer.create(
            <MemberSummaryModal memberInfo={member} memberFamily={memberFamily} memberBikes={memberBikes} />,
        ).toJSON();
        expect(modal).toMatchSnapshot();
    });
});
