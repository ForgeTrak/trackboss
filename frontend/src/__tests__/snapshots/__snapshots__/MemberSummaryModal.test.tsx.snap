// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`member summary modal renders all props correctly 1`] = `
<div>
  <button
    className="chakra-button css-1igmqd6"
    onClick={[Function]}
    type="button"
  >
    Member Summary Modal
  </button>
</div>
`;
